Implementation of the ChoiceFilter to create Elastica Filter

chervaliery opened the pull request:
For one of my project in addition to query, I need to filter my results. I have a filed 'type' in my mongodb documents and I need to search in my db and filter on one or several types.

I did not find out how to do it. But the ChoiceFilter was not implemented yet.
On the base of the StringFilter, I adapt it to make query with filters.

Furthermore, in the ElasticaDatagridBuilder, the type field of the filter was not used and you cannot choose the filter to use. I adapt it from the Sonata\DoctrineORMAdminBundle DatagridBuilder in order to use it.

greg0ire commented:
Thank you for this. I am away for two weeks so I cannot have a good look at it right now. In the meantime, could you try rewriting your commit messages according to [these rules](http://rakeroutes.com/blog/deliberate-git/)? If you don't I'll do it for you.

greg0ire commented on Builder/ElasticaDatagridBuilder.php:
```diff
@@ -26,131 +26,135 @@
+		$this->filterFactory		   = $filterFactory;
+		$this->guesser				 = $guesser;
+		$this->finderProvider		  = $finderProvider;
+		$this->configManager		   = $configManager;
```
Could you fix the formatting here ?

greg0ire commented on Builder/ElasticaDatagridBuilder.php:
```diff
@@ -26,131 +26,135 @@
-
-        return $smart;
-    }
+	private $finderProvider;
```
Looks like you replaced spaces with tabs.

chervaliery replied:
Yes, Eclipse's auto indent...
I can change my commit

greg0ire commented:
Why did you close the PR ?

chervaliery commented:
I tried to undo my commit to fix the format and commit the proper one.
But I don't know why it closed m yPR

greg0ire commented:
Ok great, I'll try to test that asap

chervaliery committed:
Implement ChoiceFilter

Adapt the ChoiceFilter from the StringFilter to create an Elasica Filter
Adapt the ElasticaDatagridBuilder in order to use the type field.

chervaliery commented:
Ok this time I think it will be good.

Sorry, I am a git/github beginner

greg0ire commented:
No problem, and thanks of lot for contributing!

greg0ire commented:
I tried your version of the bundle on my project, and immediately got this error. : 

> SearchPhaseExecutionException[Failed to execute phase [query], all shards failed; shardFailures {[YsILbfn8QA-qdlHTB08zuQ][v3_encoding_manager][0]: SearchParseException[[v3_encoding_manager][0]: from[-1],size[-1]: Parse Failure [Failed to parse source [{"sort":{"urlUpdatedAt":{"order":"desc"}},"query":{"match_all":{}}}]]]; nested: SearchParseException[[v3_encoding_manager][0]: from[-1],size[-1]: Parse Failure [No mapping found for [urlUpdatedAt] in order to sort on]]; }{[YsILbfn8QA-qdlHTB08zuQ][v3_encoding_manager][1]: SearchParseException[[v3_encoding_manager][1]: from[-1],size[-1]: Parse Failure [Failed to parse source [{"sort":{"urlUpdatedAt":{"order":"desc"}},"query":{"match_all":{}}}]]]; nested: SearchParseException[[v3_encoding_manager][1]: from[-1],size[-1]: Parse Failure [No mapping found for [urlUpdatedAt] in order to sort on]]; }{[YsILbfn8QA-qdlHTB08zuQ][v3_encoding_manager][2]: SearchParseException[[v3_encoding_manager][2]: from[-1],size[-1]: Parse Failure [Failed to parse source [{"sort":{"urlUpdatedAt":{"order":"desc"}},"query":{"match_all":{}}}]]]; nested: SearchParseException[[v3_encoding_manager][2]: from[-1],size[-1]: Parse Failure [No mapping found for [urlUpdatedAt] in order to sort on]]; }{[YsILbfn8QA-qdlHTB08zuQ][v3_encoding_manager][3]: SearchParseException[[v3_encoding_manager][3]: from[-1],size[-1]: Parse Failure [Failed to parse source [{"sort":{"urlUpdatedAt":{"order":"desc"}},"query":{"match_all":{}}}]]]; nested: SearchParseException[[v3_encoding_manager][3]: from[-1],size[-1]: Parse Failure [No mapping found for [urlUpdatedAt] in order to sort on]]; }{[YsILbfn8QA-qdlHTB08zuQ][v3_encoding_manager][4]: SearchParseException[[v3_encoding_manager][4]: from[-1],size[-1]: Parse Failure [Failed to parse source [{"sort":{"urlUpdatedAt":{"order":"desc"}},"query":{"match_all":{}}}]]]; nested: SearchParseException[[v3_encoding_manager][4]: from[-1],size[-1]: Parse Failure [No mapping found for [urlUpdatedAt] in order to sort on]]; }] 

After adding the missing mapping, it works again, but it looks like you introduced a BC break and made something that was optional mandatory…

greg0ire commented:
Also, could you give me a usage example ? I know it's funny, but I am not at all familiar with elastica, and I am having a hard time figuring out how to use the choice filter.

@clebeaupin : could you please test this ?

chervaliery commented:
Here my situation: 
I use Elastic search to index vulnerabilities stored in a mongodb database.
They have a name, a description and a type. The type can only be: cve, cwe or cpe.

My problem was to search for example through all my cve and cwe.

With only a match query, as you implement it, it was not possible. The terms query is useful to search for exact words from a list. For example:

> {
>    "terms": {
>        "type": [
>            "cve",
>            "cwe"
>        ]
>    }
> }

Will search for exactly “cve” or “cwe” in the type field so I can use a Sonata choice field type.

I hope my explications are clear enough

chervaliery commented:
About your first problem, I don't understand, how could you search on none mapping index ?
What was your situation ?

greg0ire commented:
Your explanations are very clear, thanks, I see the use case.

> About your first problem, I don't understand, how could you search on none mapping index ?

If at least one of the requested fields is not mapped, then there is a fallback to the ORM / ODM , and elasticsearch is simply not used.  Could you try to keep it that way ? Have a look at the `isSmart` method of the datagrid builder.

chervaliery commented:
I tried to add a not mapped field. If I don't use it for my research, elasticsearch is used. And if I don't, elasticsearch is not used and mongodb do all the job. It seems to works on my example :

> $datagridMapper
>    ->add('name', null, array ('label' => 'ID'))
>    ->add('summary', null, array ('label' => 'Summary'))
>    ->add('type', 'elastica_choice', array ('label' => 'Type'), 'choice', array(
>            'choices' => ['all' => 'ALL', 'cves'=> "CVE", "cpe"=>"CPE", "cwe"=>"CWE"], 'multiple' => 'True'
>    ))
>    ->add('notmappedfield',null)

chervaliery commented:
> If at least one of the requested fields is not mapped

You mean not mapped in fos_elastica but in your entity

> /*\* @ MongoDB\String(name="notmappedfield") */
> 
> protected $notmappedfield;

greg0ire commented:
Yes, that is what I mean. Maybe it is ORM-specific ? I'll try to look at it this week.

chervaliery commented:
Sorry, I do not konw, I did not try to use it with ORM, all my data is store in a MongoDB.

greg0ire commented:
I should try to add functional tests so that everyone can easily see when something is broken.

greg0ire commented:
Ok, so I cannot seem to be able to reproduce this bug on master today, I don't know why… guess it is good to merge then.
